src/System.Device.Gpio/System/Device/Gpio/Drivers/SysFsDriver.Linux.cs

```diff
@@ -290,25 +332,49 @@ protected internal override WaitForEventResult WaitForEvent(int pinNumber, PinEv
             AddPinToPoll(pinNumber, ref valueFileDescriptor, ref pollFileDescriptor, out bool closePinValueFileDescriptor);
 
             bool eventDetected = WasEventDetected(pollFileDescriptor, valueFileDescriptor, out _, cancellationToken);
+            if (_statusUpdateSleepTime > 0)
+            {
+                Thread.Sleep(_statusUpdateSleepTime); // Adding some delay to make sure that the value of the File has been updated so that we will get the right event type.
+            }
+
+            PinEventTypes detectedEventType = PinEventTypes.None;
+            if (eventDetected)
+            {
+                // This is the only case where we need to read the new state. Although there are reports of this not being 100% reliable in all situations,
+                // it seems to be working fine most of the time.
+                if (eventTypes == (PinEventTypes.Rising | PinEventTypes.Falling))
+                {
+                    detectedEventType = (Read(pinNumber) == PinValue.High) ? PinEventTypes.Rising : PinEventTypes.Falling;
+                }
+                else if (eventTypes != PinEventTypes.None)
+                {
+                    // If we're only waiting for one event type, we know which one it has to be
+                    detectedEventType = eventTypes;
+                }
+            }
 
             RemovePinFromPoll(pinNumber, ref valueFileDescriptor, ref pollFileDescriptor, closePinValueFileDescriptor, closePollFileDescriptor: true, cancelEventDetectionThread: false);
             return new WaitForEventResult
             {
                 TimedOut = !eventDetected,
-                EventTypes = eventTypes
+                EventTypes = detectedEventType,
             };
         }
 
         private void SetPinEventsToDetect(int pinNumber, PinEventTypes eventTypes)
         {
             string edgePath = Path.Combine(GpioBasePath, $"gpio{pinNumber + s_pinOffset}", "edge");
+            // Even though the pin is open, we might sometimes need to wait for access
+            SysFsHelpers.EnsureReadWriteAccessToPath(edgePath);
```

Comment:
Did you have to add this because some of our tests were failing because of this?

Reply:
Yes, I saw a few failures here.